### PR TITLE
TER-89 columns limit

### PR DESCRIPTION
**Pull request by martinjunger**
Jira: TER-89
Connection: https://github.com/keboola/connection/pull/4199

Before asking for review make sure that:

## Checklist

- [x] New client method(s) has tests
- [x] Apiary file is updated

## Release

  - [x] I gave the PR a proper label:
    * major (BC break)
    * minor (new feature)
    * patch (backwards compatible fix)
    * no release (just test changes)

### CODE REVIEW

File: tests/Backend/CommonPart1/AlterTableTest.php [hunk 1]
```diff
@@ -210,6 +210,97 @@ public function testTooManyColumns(): void
...
+                $this->getTestBucketId(),
+                $data,
+            );
+            $this->fail('There were 5000 columns man. fail.');
+        } catch (ClientException $e) {
+            $this->assertEquals('storage.tables.validation.tooManyColumns', $e->getStringCode());
+        }
+    }
+
+    public function testTooManyColumnsCountAddAnotherColumnOverLimitFailed(): void
```
Comment (jirkasemmler) [inline tests/Backend/CommonPart1/AlterTableTest.php:246]: haha, good point, to me nenapadlo, ze nemam otestovane. Ale poustel jsi to na starych backendech (=snflk, synapse... ? ). imho to ted vyhodi 500, coz nechceme. melo by to podle me vyzkouset vytvorit CollumnsCollection nez se to vubec pusti do backendu. At na to mame jednotne misto, kde to spadne.

File: tests/Backend/CommonPart1/AlterTableTest.php [hunk 2]
```diff
@@ -210,6 +210,97 @@ public function testTooManyColumns(): void
...
+            for ($i = 1; $i <= 5000; $i++) {
+                $data['columns'][] = [
+                    'name' => 'col' . $i,
+                    'definition' => [
+                        'type' => 'BYTEINT',
+                        'nullable' => true,
+                    ],
+                ];
+            }
+            $this->_client->createTableDefinition(
```
Comment (jirkasemmler) [inline tests/Backend/CommonPart1/AlterTableTest.php:236]: taky good point, prozatim se to testovalo jen na tableImport

File: tests/Backend/CommonPart1/CreateTableTest.php [hunk 3]
```diff
@@ -354,10 +354,9 @@ public function testTableCreateFromNotUploadedFileShouldThrowError(): void
     }
 
     /**
-     * @param $async
      * @dataProvider syncAsyncData
      */
-    public function testTableColumnNamesSanitize($async): void
+    public function testTableColumnNamesSanitize(bool $async): void
```
Comment (jirkasemmler) [inline tests/Backend/CommonPart1/CreateTableTest.php:359]: ❤️

File: tests/Backend/CommonPart1/AlterTableTest.php [hunk 4]
```diff
@@ -210,6 +210,97 @@ public function testTooManyColumns(): void
         }
     }
 
+    public function testTooManyColumnsCountFailed(): void
+    {
+        $this->skipTestForBackend([
+            self::BACKEND_EXASOL,
+            self::BACKEND_BIGQUERY,
+        ], 'Other backends does have other limits');
```
Comment (jirkasemmler) [inline tests/Backend/CommonPart1/AlterTableTest.php:218]:
```suggestion
        ], 'Bigquery and Exasol have no limits regarding number of rows');
```

File: tests/Backend/CommonPart1/AlterTableTest.php [hunk 5]
```diff
@@ -210,6 +210,97 @@ public function testTooManyColumns(): void
...
+                        'type' => 'BYTEINT',
+                        'nullable' => true,
+                    ],
+                ];
+            }
+            $this->_client->createTableDefinition(
+                $this->getTestBucketId(),
+                $data,
+            );
+            $this->fail('There were 5000 columns man. fail.');
```
Comment (jirkasemmler) [inline tests/Backend/CommonPart1/AlterTableTest.php:240]: co znamena to `man.` ?

File: tests/Backend/CommonPart1/AlterTableTest.php [hunk 6]
```diff
@@ -210,6 +210,97 @@ public function testTooManyColumns(): void
...
 
+    public function testTooManyColumnsCountFailed(): void
+    {
+        $this->skipTestForBackend([
+            self::BACKEND_EXASOL,
+            self::BACKEND_BIGQUERY,
+        ], 'Other backends does have other limits');
+
+        try {
+            $data = [
```
Comment (jirkasemmler) [inline tests/Backend/CommonPart1/AlterTableTest.php:221]: to generovani v try byt nemusi

File: tests/Backend/CommonPart1/AlterTableTest.php [hunk 7]
```diff
@@ -210,6 +210,97 @@ public function testTooManyColumns(): void
...
+            $this->assertEquals('storage.tables.validation.tooManyColumns', $e->getStringCode());
+        }
+    }
+
+    public function testTooManyColumnsCountAddAnotherColumnOverLimitFailed(): void
+    {
+        $this->skipTestForBackend([
+            self::BACKEND_EXASOL,
+            self::BACKEND_BIGQUERY,
+        ], 'Other backends does have other limits');
```
Comment (jirkasemmler) [inline tests/Backend/CommonPart1/AlterTableTest.php:251]:
```suggestion
        ], 'Bigquery and Exasol have no limits regarding number of rows');
```

**Review by jirkasemmler — CHANGES_REQUESTED** [review]
jen par malickosti. Logika testu mi prijde fajn.

File: tests/Backend/CommonPart1/AlterTableTest.php (hunk 5 above)
Comment (martinjunger) [reply tests/Backend/CommonPart1/AlterTableTest.php:240]: tos tam napsal v cervenci ty :-) Chvili jsem nad tim dumal, ale nejspis to bude "je tam 5000 sloupcu kámo. spadlo to."

Comment (jirkasemmler) [reply tests/Backend/CommonPart1/AlterTableTest.php:240]: haha, tak z toho prosim udelej neco smysluplneho.

File: tests/Backend/CommonPart1/AlterTableTest.php (hunk 4 above)
Comment (martinjunger) [reply tests/Backend/CommonPart1/AlterTableTest.php:218]: upraveno

File: tests/Backend/CommonPart1/AlterTableTest.php (hunk 5 above)
Comment (martinjunger) [reply tests/Backend/CommonPart1/AlterTableTest.php:240]: upraveno

File: tests/Backend/CommonPart1/AlterTableTest.php (hunk 7 above)
Comment (martinjunger) [reply tests/Backend/CommonPart1/AlterTableTest.php:251]: upraveno

File: tests/Backend/CommonPart1/AlterTableTest.php (hunk 6 above)
Comment (martinjunger) [reply tests/Backend/CommonPart1/AlterTableTest.php:221]: upraveno

**Comment by martinjunger** [comment]
Odebral jsem test pro `addColumn` a vrazil to do [CT-862](https://keboola.atlassian.net/browse/CT-862).

File: tests/Backend/CommonPart1/AlterTableTest.php [hunk 8]
```diff
@@ -204,7 +204,38 @@ public function testTooManyColumns(): void
...
+        // generate table definition with too many columns - over limit on every backend
+        // don't make that number ridiculously big. RequestObject won't make it...
+        for ($i = 1; $i <= 2100; $i++) {
+            $data['columns'][] = [
+                'name' => 'col' . $i,
+                'basetype' => 'BOOLEAN',
+            ];
+        }
+        try {
+            $this->_client->createTableDefinition(
```
Comment (jirkasemmler) [inline tests/Backend/CommonPart1/AlterTableTest.php:234]: nerikal jsi, ze toto presunes nekam jinam, kde maji projekty (i synapse nebo snflk) tables-definition feature? poznas to podle init commandu v connection

**Review by jirkasemmler — CHANGES_REQUESTED** [review]
prosim presun ten test at to ma vzdycky tu spravnou feature a netestuje se to proti RS

**Comment by martinjunger** [comment]
- [ ] presunout test `AlterTableTest::testTooManyColumnsCountFailed`
  - v `paratest-backend-snowflake-gcs-part-1` nema spravnou feature ... `projectMissingFeatureFlag`
  - mozna i v `paratest-backend-snowflake-part-1` a `paratest-backend-snowflake-abs-part-1` a `paratest-backend-synapse-part-1` ... `accessDenied`

File: tests/Backend/CommonPart1/AlterTableTest.php (hunk 8 above)
Comment (martinjunger) [reply tests/Backend/CommonPart1/AlterTableTest.php:234]: jj, pridal jsem komentar puvodne do Connection PR, ted uz je tady...

**Comment by martinjunger** [comment]
> prosim presun ten test at to ma vzdycky tu spravnou feature a netestuje se to proti RS

Presunuto do CommonPart2, kde jsou i testy pro `tables-definition` v `CommonPart2`.

**Review by jirkasemmler — APPROVED** [review]

**Comment by martinjunger** [comment]
🏷️ [Tag 1](https://dev.azure.com/keboola-dev/Connection%20build/_build/results?buildId=30877&view=results) bezel a skoncil 🔴  - jdu to prozkoumat...

**Pull request merged**